### PR TITLE
Extended Yara match structures, @Extractor.string with additional `match` argument

### LINKED ISSUE #5: Provide access to the content of the matched strings
**Description**

Instead of only providing `p` and `offset`, it would be great to get the content of the matched string from Yara. This way, we don't have to guess the length of the matched string (for example, when dealing with wildcarded regex)

I find it unnecessary complicated to not have access to the content of the matched strings, both in `Extractor` string decorators and in the `handle_yara` function.

To demonstrate, here is a simple output from Yara. You can see that the output contains `OFFSET : MATCHED STRING : CONTENT`
```
yara -s test.yara tests/samples/e6c73621b9718d187ad809c8f2e94953e2e702fd43021bc8c3966f7698165491


elknot tests/samples/e6c73621b9718d187ad809c8f2e94953e2e702fd43021bc8c3966f7698165491

0xf5d8b:$reg1: \x00206-37/133/18\x00\x00\x00\x00\x00

0x1d8b57:$reg1: \x00277-3-256-37\x00\x00\x00\x00\x00

0x1d8b65:$reg2: \x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x002/862\x00

```

### PR DESCRIPTION
**Pull request by psrok1**
- Extended version of `@Extractor.extractor` methods called `@Extractor.string`
```python
class Evil(Extractor):
    yara_rules = ("evil", "weird")
    family = "evil"
    ...

    @Extractor.string("accepts", "multiple", "strings")
    def hello_evil(self, p: ProcessMemory, addr: int, match: YaraStringMatch):
        # Accepts multiple identifiers, called by $accepts, $multiple and $strings hit
        if match.identifier == "accepts":
            # Allows to get both hit VA and matched content
            return {"hello_message": match.content}
```

- `malduck.yara` extended structures: `YaraRulesetMatch`, `YaraRuleMatch`, `YaraStringMatch`
- structures offering legacy offset-only interface: `YaraRulesetOffsets` (`YaraMatches`), `YaraRuleOffsets` (`YaraMatch`)
- both extended and legacy versions are available via `ProcessMemory.yarap/v` and `Yara.match` methods. We can choose new structures via additional `extended=True` argument, which is set to False by default.

Other changes:
- `YaraRulesetMatch.remap` doesn't mutate the object, returns remapped object instead
- Refactor of Yara match mapping machinery

closes #5 

Missing parts:
- [x] Documentation
- [x] Tests for new features

### CODE REVIEW

**Comment by ITAYC0HEN** [comment]
< asking myself to review >

File: malduck/yara.pyi [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+class YaraRulesetOffsets(_Mapper["YaraRuleOffsets"]):
+    _matches: YaraRulesetMatch
+    def __init__(self, matches: YaraRulesetMatch) -> None:
+        super().__init__(elements={})
+    def remap(
+        self, offset_mapper: Optional[OffsetMapper] = None
+    ) -> "YaraRulesetOffsets": ...
+
+YaraStringMatch = namedtuple("YaraStringMatch", ["identifier", "hit", "content"])
```
Comment (ITAYC0HEN) [inline malduck/yara.pyi:123]: I don't like the naming `hit` in compared to `offset`.

Id prefer to access the address in which the string found as in `match.offset` and not `match.hit`

Comment (psrok1) [reply malduck/yara.pyi:123]: Agree, it's inconsistent with all these "offsets" and "offset_mappers" around. To be renamed.

**Comment by psrok1** [comment]
Tested on CERT.pl extraction unit tests: compatibility check passed :heavy_check_mark: (with some obvious `Extractor.handle_yara` deprecation warnings)

File: malduck/procmem/procmem.py [hunk 2]
```diff
@@ -766,18 +781,19 @@ def map_offset(off, len):
...
+            offset_mapper=map_offset, extended=extended, data=self.readp(0)
+        )
 
     def _findbytes(self, yara_fn, query, addr, length):
         if isinstance(query, bytes):
             query = query.decode()
 
         rule = Yara(strings=YaraString(query, type=YaraString.HEX))
-        match = yara_fn(rule, addr, length)
+        match = yara_fn(rule, addr, length, True)
```
Comment (nazywam) [inline malduck/procmem/procmem.py:793]: This free-hanging `True` looks really weird, are we allowed to use kwargs here or not?

File: malduck/yara.py [hunk 3]
```diff
@@ -137,10 +185,19 @@ def match(self, offset_mapper=None, **kwargs):
         :type offset_mapper: function
```
[the remark this answers is not in the document]
Comment (psrok1) [reply malduck/yara.py:185]: Fixed

File: malduck/procmem/procmem.py (hunk 2 above)
Comment (psrok1) [reply malduck/procmem/procmem.py]: Fixed

**Pull request merged**